### PR TITLE
[tune] Fix local checkpoint deletion for remote trials

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

The changes in #18318 introduced a regression for non-durable trainables - checkpoint copies on the driver were not removed anymore, as the runner ip always equaled the node ip. 

Instead we can fix both problems by simply calling the remote delete function before we check for a local copy. This will work for both durable and non-durable trainables.

## Related issue number

This PR partly reverts #18318

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by xwjiang2010 — APPROVED** [review]
lgtm. Just left a question on why deleting checkpoint on cloud storage relies on some directory structure on the Trainable node.

**Review by Yard1 — APPROVED** [review]

File: python/ray/tune/trial.py [hunk 1]
```diff
@@ -117,22 +99,25 @@ def __call__(self, checkpoint):
...
-                         checkpoint.value)
             checkpoint_path = checkpoint.value
 
+            logger.debug("Trial %s: Deleting checkpoint %s", self.trial_id,
+                         checkpoint_path)
+
+            # TODO(ujvl): Batch remote deletes.
+            # We first delete the remote checkpoint. If it is on the same
+            # node as the driver, it will also remove the local copy.
+            ray.get(self.runner.delete_checkpoint.remote(checkpoint_path))
```
[the remark this answers is not in the document]
Comment (jovany-wang) [reply python/ray/tune/trial.py:110]: Nice fix! We had this fix for our internal version already.

[the remark this answers is not in the document]
Comment (jovany-wang) [reply python/ray/tune/trial.py:110]: >  [trainable.delete_checkpoint.remote(checkpoint_path), delete_checkpoint_on_driver(checkpoint_path)], and the two should not be interfering with each other.

This approach looks better to me!

[the remark this answers is not in the document]
Comment (jovany-wang) [reply python/ray/tune/trial.py:110]: @xwjiang2010 @krfricke Are you guys willing to make the two deletions in parallel?

**Pull request merged**